_posts/2019-03-26-read-models-a-knot-to-untie.md

```diff
@@ -0,0 +1,102 @@
+
+![Related image](https://i.stack.imgur.com/eKU6r.png){:class="img-responsive"}
+
+The majority of the described implementations imply that there is some sort of synchronization between the write storage, `RDBMS` in the image, and the read storage, the incredibly fast thing.
+
+There is really nothing preventing us to sync in some way data from different services to the incredibly fast read storage. For example:
+
+- *Marketing* sends name and description.
+- *Sales* sends price.
+- *Warehouse* sends availability.
+- *Shipping* sends shipping options.
+
+Given that all services share the same identifier, a _Product_ read model can be crafted and stored.
+
+### Keep It Simple Stupid
+
+Is there a simpler solution? So far we've used code snippets like:
+
+```csharp
+class SampleHandler : IHandleRequests
+{
+   public bool Matches(RouteData routeData, string httpVerb, HttpRequest request)
+   {
+      /* omitted for clarity */
+   }
+   
+   public async Task Handle(dynamic vm, RouteData routeData, HttpRequest request)
+   {
+      /* omitted for clarity */
+   }
+}
+```
+
+the infrastructure executing the above code could, at runtime, store the resulting ViewModel, the `dynamic vm`, in a dedicated incredibly fast storage, and query that storage for all the subsequent requests for the same URL. We've basically created a read model at runtime at the first request. So far, so good.
+
+### You know what?
+
+*That's a cache*. It's the sad truth, but we should be honest with ourselves: what we just designed is a cache.
+
+## Ownership dilemma
+
+As soon as the solution is identified for what it is, and especially given the distributed system context, it immediately shines as something we probably don't want.
+
+> Caching comes with its own set of issues, distributed caching is far worse.
+
+One the biggest architectural issues that comes along with caching composed read models is that there is no service owning it anymore.
+
+### Businesses change their mind
+
+A business requirement in _Marketing_ could change the way data should be presented to users, in essence invalidating all the cached read models. _Marketing_ doesn't own the resulting model, thus the only option we're left with is to invalidate the whole set. _Product_ read models are wiped from the "cache" and:
+
+- in the first scenario, the fancy diagram, we end up with an incredibly fast but empty read storage. storage that needs to be re-synchronized, with the risk of huge delays
+- in the second scenario, the cached ViewModel, we end with an empty cache that will be gradually rebuilt causing a huge load on the whole system
+
+In both scenarios we'll end up with a huge load on the entire system. All services will be hammered, in one way or the other, for the sake of nothing since the only ting that changed is _Marketing_.
```

Comment:
```suggestion
In both scenarios we'll end up with a huge load on the entire system. All services will be hammered, in one way or the other, for the sake of nothing since the only thing that changed is _Marketing_.
```